implement reindex using an array of labels

```python
>>> arr = ndrange('a=v0..v1;b=v0..v2')
>>> arr
a\b  v0  v1  v2
 v0   0   1   2
 v1   3   4   5
>>> arr.reindex('a', arr.b.labels)
-> 1349             assert isinstance(value, Axis)
AssertionError: 
```

--- a/doc/source/changes/version_0_26.rst.inc
+++ b/doc/source/changes/version_0_26.rst.inc
@@ -12,6 +12,31 @@ Miscellaneous improvements
 
 * view() and edit() without argument now display global arrays in addition to local ones (closes :editor_issue:`54`).
 
+* allowed to pass an array of labels as `new_axis` argument to `reindex` method (closes :issue:`384`):
+
+    >>> arr = ndrange('a=v0..v1;b=v0..v2')
+    >>> arr
+    a\b  v0  v1  v2
+     v0   0   1   2
+     v1   3   4   5
+    >>> arr.reindex('a', arr.b.labels)
+    a\b   v0   v1   v2
+     v0    0    1    2
+     v1    3    4    5
+     v2  nan  nan  nan
+
+* allowed to call the `reindex` method using a differently named axis for labels (closes :issue:`386`):
+
+    >>> arr = ndrange('a=v0..v1;b=v0..v2')
+    >>> arr
+    a\b  v0  v1  v2
+     v0   0   1   2
+     v1   3   4   5
+    >>> arr.reindex('a', arr.b)
+    a\b   v0   v1   v2
+     v0    0    1    2
+     v1    3    4    5
+     v2  nan  nan  nan
 
 Fixes
 -----

--- a/larray/core/array.py
+++ b/larray/core/array.py
@@ -1342,7 +1342,7 @@ class LArray(ABCLArray):
             Axes to reindex. If a single axis reference is given, the `new_axis` argument must be provided.
             If a list of Axis or an AxisCollection is given, all axes will be reindexed by the new ones.
             In that case, the number of new axes must match the number of the old ones.
-        new_axis : int, str, list/tuple of str or Axis, optional
+        new_axis : int, str, list/tuple/array of str or Axis, optional
             List of new labels or new axis if `axes_to_replace` contains a single axis reference.
         fill_value : scalar or LArray, optional
             Value set to data corresponding to added labels. Defaults to NaN.
@@ -1398,7 +1398,7 @@ class LArray(ABCLArray):
 
         Reindex using axes from another array
 
-        >>> arr2 = ndrange('a=a0..a1;c=c0..c0;b=b0..b2')
+        >>> arr2 = ndrange('a=a0,a1;c=c0..c0;b=b0..b2')
         >>> arr2
          a  c\\b  b0  b1  b2
         a0   c0   0   1   2
@@ -1417,8 +1417,10 @@ class LArray(ABCLArray):
         a1   c0  3.0  4.0
         """
         # XXX: can't we move this to AxisCollection.replace?
-        if isinstance(new_axis, (int, basestring, list, tuple)):
+        if isinstance(new_axis, (int, basestring, list, tuple, np.ndarray)):
             new_axis = Axis(new_axis, self.axes[axes_to_reindex].name)
+        if isinstance(new_axis, Axis):
+            new_axis = new_axis.rename(self.axes[axes_to_reindex].name)
         if isinstance(axes_to_reindex, AxisCollection):
             assert new_axis is None
             # add extra axes if needed

--- a/larray/tests/test_array.py
+++ b/larray/tests/test_array.py
@@ -2382,6 +2382,21 @@ age    0       1       2       3       4       5       6       7        8  ...  
                                                  a0   1   0   0
                                                  a1   3   1   2"""))
 
+        # using labels from another array
+        arr = ndrange('a=v0..v2;b=v0,v2,v1,v3')
+        res = arr.reindex('a', arr.b.labels, fill_value=-1)
+        assert_array_equal(res, from_string("""a\\b  v0  v2  v1  v3
+                                                 v0   0   1   2   3
+                                                 v2   8   9  10  11
+                                                 v1   4   5   6   7
+                                                 v3  -1  -1  -1  -1"""))
+        res = arr.reindex('a', arr.b, fill_value=-1)
+        assert_array_equal(res, from_string("""a\\b  v0  v2  v1  v3
+                                                 v0   0   1   2   3
+                                                 v2   8   9  10  11
+                                                 v1   4   5   6   7
+                                                 v3  -1  -1  -1  -1"""))
+
     def test_append(self):
         la = self.small
         sex, lipro = la.axes

